src/lib/storage/index/partial_hash/partial_hash_index_impl.cpp

```diff
@@ -0,0 +1,165 @@
+#include "partial_hash_index_impl.hpp"
+
+#include "storage/segment_iterate.hpp"
+
+namespace hyrise {
+
+template <typename DataType>
+PartialHashIndexImpl<DataType>::PartialHashIndexImpl(
+    const std::vector<std::pair<ChunkID, std::shared_ptr<Chunk>>>& chunks_to_index, const ColumnID column_id)
+    : BasePartialHashIndexImpl() {
+  insert_entries(chunks_to_index, column_id);
+}
+
+template <typename DataType>
+size_t PartialHashIndexImpl<DataType>::insert_entries(
+    const std::vector<std::pair<ChunkID, std::shared_ptr<Chunk>>>& chunks_to_index, const ColumnID column_id) {
+  auto indexed_chunks = size_t{0};
+  for (const auto& chunk : chunks_to_index) {
+    if (_indexed_chunk_ids.contains(chunk.first)) {
+      // Index already contains entries for the given chunk.
+      continue;
+    }
+
+    _indexed_chunk_ids.insert(chunk.first);
+    ++indexed_chunks;
+
+    // Iterate over the segment to index and populate the index.
+    const auto& indexed_segment = chunk.second->get_segment(column_id);
+    segment_iterate<DataType>(*indexed_segment, [&](const auto& position) {
+      const auto row_id = RowID{chunk.first, position.chunk_offset()};
+      // If value is NULL, add to NULL vector, otherwise add into value map.
+      if (position.is_null()) {
+        if (!_null_positions.contains(DataType{})) {
+          _null_positions[DataType{}] = std::vector<RowID>();
+        }
+        _null_positions[DataType{}].push_back(row_id);
+      } else {
+        _positions[position.value()].emplace_back(row_id);
+      }
+    });
+  }
+
+  return indexed_chunks;
+}
+
+template <typename DataType>
+size_t PartialHashIndexImpl<DataType>::remove_entries(const std::vector<ChunkID>& chunks_to_unindex) {
+  const size_t size_before = _indexed_chunk_ids.size();
+
+  auto indexed_chunks_to_unindex = std::unordered_set<ChunkID>{};
+  for (const auto& chunk_id : chunks_to_unindex) {
+    if (!_indexed_chunk_ids.contains(chunk_id)) {
+      continue;
+    }
+
+    indexed_chunks_to_unindex.insert(chunk_id);
+    _indexed_chunk_ids.erase(chunk_id);
+  }
+
+  // Checks whether a given RowID's ChunkID is in the set of ChunkIDs to be unindexed.
+  auto is_to_unindex = [&indexed_chunks_to_unindex](const RowID& row_id) {
+    const auto find_iter =
+        std::find(indexed_chunks_to_unindex.cbegin(), indexed_chunks_to_unindex.cend(), row_id.chunk_id);
```

Comment:
You use std::find to check if `row_id.chunk_id` is contained in `std::unordered_set<ChunkID>{}`, right?
`return indexed_chunks_to_unindex.contains(row_id.chunk_id)`?